internal/runner/operation_project_destroy.go

```diff
@@ -0,0 +1,132 @@
+package runner
+
+import (
+	"context"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/waypoint/internal/core"
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+	empty "google.golang.org/protobuf/types/known/emptypb"
+)
+
+func (r *Runner) executeDestroyProjectOp(
+	ctx context.Context,
+	log hclog.Logger,
+	job *pb.Job,
+	project *core.Project,
+) (*pb.Job_Result, error) {
+	client := project.Client()
+	destroyProjectOp, ok := job.Operation.(*pb.Job_DestroyProject)
+	if !ok {
+		// this shouldn't happen since the call to this function is gated
+		// on the above type match.
+		panic("operation not expected type")
+	}
+
+	// Update the project state to indicate that it's being destroyed
+	pbProject, err := client.UpsertProject(ctx,
+		&pb.UpsertProjectRequest{
+			Project: &pb.Project{
+				Name:  destroyProjectOp.DestroyProject.Project.Name,
+				State: pb.Project_DESTROYING,
+			},
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if !destroyProjectOp.DestroyProject.SkipDestroyResources {
+		for _, app := range project.Apps() {
+			// Get the workspaces for the app we're currently destroying
+			workspaces, err := client.ListWorkspaces(ctx, &pb.ListWorkspacesRequest{
+				Scope: &pb.ListWorkspacesRequest_Application{
+					Application: &pb.Ref_Application{
+						Project:     destroyProjectOp.DestroyProject.Project.Name,
+						Application: app,
+					},
+				},
+			})
+			if err != nil {
+				return nil, err
+			}
+			// Destroy the resources in the workspaces for the app
+			for _, workspace := range workspaces.Workspaces {
+				log.Debug("Destroying resources in workspace %s", workspace.Name)
+				// TODO: The destroy operation doesn't currently respect the Workspace
+				// being set on the Job - this should be updated
```

Comment:
It would be nice to fix this! The behavior seems to be rooted in [runner/operation.go](https://github.com/hashicorp/waypoint/blob/main/internal/runner/operation.go#L292-L303). The workspace being used in the destroyment is derived from the `core.Project` that is set here:

```go
	project, err := core.NewProject(ctx,
		core.WithLogger(log),
		core.WithUI(ui),
		core.WithComponents(factories),
		core.WithClient(r.client),
		core.WithConfig(cfg),
		core.WithDataDir(projDir),
		core.WithLabels(job.Labels),
		core.WithVariables(inputVars),
		core.WithWorkspace(job.Workspace.Workspace),  <--- here
		core.WithJobInfo(jobInfo),
	)
// some other code
	case *pb.Job_DestroyProject:
		return r.executeDestroyProjectOp(ctx, log, job, project)
```

So in a way, my comment isn't accurate - the destroyment ONLY respects the workspace that is set on the job and no others... but in our case that is the project destroy job, which happens to need to loop through all workspaces and delete everything. What this means is that if you ran `project destroy -w=dev`, then all of the deployments in a `dev` workspace _would_ be deleted, but no others. But that top-level workspace flag controls which workspace's deployments are going to be destroyed... need to think more on this one but I think this defines the issue.